File: daily/Binary Tree/110-Balanced-binary-tree.py
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(node):
            if not node:
                return 0, True
            leftHeight, leftBalance = dfs(node.left)
            rightHeight, rightBalance = dfs(node.right)

            currentBalance = leftBalance and rightBalance and abs(leftHeight - rightHeight) <= 1
            currentHeight = 1 + max(leftHeight, rightHeight)
            return (currentHeight, currentBalance)
        return dfs(root)[1]

File: daily/Binary Tree/110-Balanced-binary-tree.go
package daily

func isBalanced(root *TreeNode) bool {
	var dfs func(*TreeNode) (int, bool)
	dfs = func(node *TreeNode) (int, bool) {
		if node == nil {
			return 0, true
		}
		leftHeight, leftBalance := dfs(node.Left)
		rightHeight, rightBalance := dfs(node.Right)

		diff := leftHeight - rightHeight
		if diff < 0 {
			diff = -diff
		}
		currentBalance := leftBalance && rightBalance && diff <= 1
		currentHeight := 1 + max(leftHeight, rightHeight)
		return currentHeight, currentBalance
	}
	_, balanced := dfs(root)
	return balanced
}
